### PR TITLE
feat: support for new enrichment logic in traces

### PR DESCRIPTION
**Pull request by nityanandagohain**
Support for new enrichment logic.
* Support for multiple type/datatype as in logs.
* Support for the new and old formats. Old format and static ones are served though the map in the constants file.
* serviceName is replaced with service.Name in filters.

Part of https://github.com/SigNoz/signoz/issues/5713

---
Missing scope:- 

* There will be a supporting PR after this which will take care of getting all the keys from the DB as truth with the context that if it is materialized or not, attribute/resource and the corresponding datatype.
* V2 of function in clickhouse reader that uses the new table.
* Integration of the flag to use the new enrichment logic and the query builder.
   * every place the builder is used.

----

In future we can try to merge the enrich logic of logs and traces but for now I am keeping it separete.

----

> [!IMPORTANT]
> Adds new enrichment logic for traces, supporting multiple data types and formats, with updated tests and constants.
> 
>   - **Behavior**:
>     - Adds new enrichment logic for traces in `enrich.go`, supporting multiple data types and formats.
>     - Replaces `serviceName` with `service.name` in trace filters.
>   - **Functions**:
>     - Introduces `EnrichTracesQuery()` and `enrichKeyWithMetadata()` in `enrich.go` for trace enrichment.
>     - Modifies `GenerateLogEnrichmentKeys()` to `GenerateEnrichmentKeys()` in `logs.go`.
>   - **Constants**:
>     - Updates `StaticFieldsTraces` in `constants.go` with new fields like `response_status_code` and `external_http_url`.
>   - **Tests**:
>     - Adds `enrich_test.go` for testing trace enrichment logic.
>     - Updates `query_builder_test.go` with new test cases for trace queries.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=SigNoz%2Fsignoz&utm_source=github&utm_medium=referral)<sup> for 87a4b973ac66f86e59abcc47482a5c0d9ffe11f3. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: pkg/query-service/app/traces/v4/enrich.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	v3 "go.signoz.io/signoz/pkg/query-service/model/v3"
+	"go.signoz.io/signoz/pkg/query-service/utils"
+)
+
+// if the field is timestamp/id/value we don't need to enrich
+// if the field is static we don't need to enrich
+// for all others we need to enrich
+// an attribute/resource can be materialized/dematerialized
+// but the query should work regardless and shouldn't fail
+func isEnriched(field v3.AttributeKey) bool {
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply pkg/query-service/app/traces/v4/enrich.go:14]: duplicate is fine for now

**Comment by srikanthccv** [comment]
@nityanandagohain, we should have the table schema-based logic in the traces as well. For example, if I create a `resource_string_service$$namepace` (doesn't matter how it's created assume it is created) then `service.namespace` should automatically routed to the column on the table. I don't see it in this PR.

**Comment by srikanthccv** [comment]
>If i have a key named service.namespace in resource attributes and there is a mat column resource_string_service$$namesapce, then the keys map will contains


Where exactly does this happen? How does it know there is a mat column `resource_string_service$$namesapce`?

**Comment by nityanandagohain** [comment]
If you see the logs existing code, there is a function named `GetLogFields` which we call before enrichment of logs . There we get all the fields which have a corresponding materialized column. So this enrichment functions will already have the context if there exists a materialized column for corresponding attribute. That's what we pass to the Enrich() function as second parameter.

**Comment by srikanthccv** [comment]
In case my question is not clear, I am asking where is it done for traces https://github.com/SigNoz/signoz/pull/6438#issuecomment-2478177158?

**Comment by nityanandagohain** [comment]
As I mentioned, it will be added in further PR's that integrates this Enrichment function. And there the logic for getting the `keys` which is a parameter for this enrichment function will be added..

**Comment by srikanthccv** [comment]
bear with me; please help me understand the breakdown. What is done in this PR, and what is coming in the next set of PRs and what are those PRs? I could neither understand any of that from the PR description nor is it highlighted in the linked issue.

**Comment by nityanandagohain** [comment]
have updated the description with the next set of things that are coming.

**Comment by srikanthccv** [comment]
Please fix the build-pipeline

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**